### PR TITLE
refactor(swc_core): enable bytechecked flag

### PR DESCRIPTION
**Pull request by kwonoj**
**Description:**

Another update to enable bytecheck rkyv feature flag.

Our dep tree is quite tangled with implicit enable (plugin -> rkyv). PR tries to detach some of it while trying to preserve existing behavior as much as it can.

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

**Review by kdy1 — COMMENTED** [review]
Did you verify rustdoc action, too?

**Comment by kwonoj** [comment]
Do you mean this? https://github.com/swc-project/swc/blob/54b108dc5133e27d8b4c48b35892a7928174fce7/.github/workflows/docs.yml#L25 at least it seems to build without failure.

File: crates/swc_common/src/plugin/mod.rs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 pub mod diagnostics;
 pub mod metadata;
-#[cfg(feature = "plugin-base")]
-#[cfg_attr(docsrs, doc(cfg(feature = "plugin-base")))]
+#[cfg(feature = "__plugin")]
```
Comment (kdy1) [inline crates/swc_common/src/plugin/mod.rs:3]: I think this name is misleading

Comment (kwonoj) [reply crates/swc_common/src/plugin/mod.rs:3]: I feel overall it's a bit complex with namings, while I choose the short path as I think we may get rid of all of this once verification is done. If we're going to keep this permanent, it'd be probably better to organize bit more.

File: crates/swc_plugin_runner/src/lib.rs [hunk 2]
```diff
@@ -7,8 +7,11 @@ use transform_executor::TransformExecutor;
 
 pub mod cache;
 mod host_environment;
+#[cfg(feature = "__rkyv")]
```
Comment (kdy1) [inline crates/swc_plugin_runner/src/lib.rs:10]: I think it's better to have

```rust
#[cfg(all(not(feature = "rkyv"), not(feature = "rkyv-bytecheck-impl")))
compile_error!("Error message")
```

Comment (kdy1) [reply crates/swc_plugin_runner/src/lib.rs:10]: `swc_plugin_*` is useless if there's no rkyv, and we should fail on compile time instead of runtime

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

File: crates/swc_plugin_runner/src/lib.rs (hunk 2 above)
Comment (kwonoj) [reply crates/swc_plugin_runner/src/lib.rs:10]: There's no fancy message, but this still fails on compile time if any plugin transform misses rkyv-impl or rkyv-bytecheck-impl.

Comment (kwonoj) [reply crates/swc_plugin_runner/src/lib.rs:10]: after all, I'm not too much worried about those case as `swc_core` should have right control for those, between `plugin_transform` vs `plugin_transform_bytecheck`.

File: crates/swc_plugin_runner/src/lib.rs [hunk 3]
```diff
@@ -29,3 +33,14 @@ pub fn create_plugin_transform_executor(
...
+
+#[cfg(not(feature = "__rkyv"))]
+pub fn create_plugin_transform_executor(
+    path: &Path,
+    cache: &Lazy<cache::PluginModuleCache>,
+    source_map: &Arc<SourceMap>,
+    metadata_context: &Arc<TransformPluginMetadataContext>,
+    plugin_config: Option<serde_json::Value>,
+) -> Result<TransformExecutor, Error> {
+    unimplemented!("Transform plugin cannot be used without serialization support")
```
Comment (kdy1) [inline crates/swc_plugin_runner/src/lib.rs:45]: I was worried about this

Comment (kdy1) [reply crates/swc_plugin_runner/src/lib.rs:45]: Is this really required?
I think it should not compile if there's no rkyv

Comment (kwonoj) [reply crates/swc_plugin_runner/src/lib.rs:45]: all of the imports are guarded by flags, this won't occur in normal condition. This was introduced to let plain `cargo check` does not fail due to missing exported signatures.

I'd like to remove it, but then all of CI / and `cargo check` will break right away.

**Review by kdy1 — APPROVED** [review]
Thanks!

---

swc-bump:
 - swc_atoms
 - swc_common --breaking

**Pull request merged**